test/quickfort/ecosystem.lua

```diff
@@ -0,0 +1,291 @@
+    -- read in test plan
+    local sets = get_blueprint_sets()
+
+    local area = {width=0, height=0, depth=0}
+    for basename,set in pairs(sets) do
+        print(('running quickfort test: "%s": %s'):
+                    format(basename, set.spec.description))
+
+        -- find an unused area of the map that meets requirements, else skip
+        if not get_test_area(area, set.spec) then
+            print(('cannot find unused map area to test set "%s"; skipping'):
+                  format(basename))
+            goto continue
+        end
+
+        -- quickfort run #dig blueprint (or just designate the whole block if
+        -- there is no #dig blueprint)
+        local modes = set.modes
+        if modes.dig then
+            quickfort_run(modes.dig.listnum, area.pos)
+        else
+            designate_area(area.pos, set.spec)
+        end
+
+        -- run dig-now to dig out designated tiles
+        dfhack.run_command('dig-now')
+
+        -- quickfort run remaining blueprints
+        for _,mode_name in pairs(mode_names) do
+            if mode_name ~= 'dig' and modes[mode_name] then
+                quickfort_run(modes[mode_name].listnum, area.pos)
+            end
+        end
+
+        -- run blueprint to generate files in output dir
+        run_blueprint(basename, set, area.pos)
+
+        -- quickfort undo blueprints
+        for _,mode_name in pairs(mode_names) do
+            if modes[mode_name] then
+                quickfort_undo(modes[mode_name].listnum, area.pos)
+            end
+        end
+
+        -- run tiletypes to reset tiles in area to hidden walls
+        reset_area(area, set.spec)
+
+        -- compare md5sum of input and output files
+        local md5File = dfhack.internal.md5File
+        for mode,mode_data in pairs(modes) do
+            if mode == 'notes' then goto continue end
+            local input_filepath = mode_data.input_filepath
+            local output_filepath = mode_data.output_filepath
+            local input_hash, input_size = md5File(input_filepath)
+            local output_hash, output_size = md5File(output_filepath)
+            expect.eq(input_hash, output_hash,
+                      'compare blueprint contents to input: '..output_filepath)
+            expect.eq(input_size, output_size,
+                      'compare blueprint length to input: '..output_filepath)
+            if not output_hash then goto continue end
+            if input_hash ~= output_hash or input_size ~= output_size then
+                -- show diff
+                local input, output =
+                    io.open(input_filepath, 'r'), io.open(output_filepath, 'r')
+                local input_lines, output_lines = {}, {}
+                for l in input:lines() do table.insert(input_lines, l) end
+                for l in output:lines() do table.insert(output_lines, l) end
+                input:close()
+                output:close()
+                expect.table_eq(input_lines, output_lines)
```

Comment:
This file-comparing logic _might_ be useful to move into `expect.lua`, although I don't expect (heh) to have many tests that compare files, so it can wait for now.